diff --git a/gdk/broadway/broadway-protocol.h b/gdk/broadway/broadway-protocol.h
index a343617..c098fdf 100644
--- a/gdk/broadway/broadway-protocol.h
+++ b/gdk/broadway/broadway-protocol.h
@@ -160,7 +160,7 @@ typedef struct {
 typedef struct {
   BroadwayRequestBase base;
   guint32 id;
-  char name[34];
+  char name[36];
   guint32 width;
   guint32 height;
 } BroadwayRequestUpdate;
diff --git a/gdk/broadway/broadwayd.c b/gdk/broadway/broadwayd.c
index 9a273c7..1c06eb7 100644
--- a/gdk/broadway/broadwayd.c
+++ b/gdk/broadway/broadwayd.c
@@ -409,7 +409,6 @@ client_fill_cb (GObject *source_object,
       guint32 size;
       gsize count, remaining;
       guint8 *buffer;
-      BroadwayRequest request;
 
       buffer = (guint8 *)g_buffered_input_stream_peek_buffer (client->in, &count);
 
@@ -417,14 +416,10 @@ client_fill_cb (GObject *source_object,
       while (remaining >= sizeof (guint32))
 	{
 	  memcpy (&size, buffer, sizeof (guint32));
-	  
+
 	  if (size <= remaining)
 	    {
-	      g_assert (size >= sizeof (BroadwayRequestBase));
-	      g_assert (size <= sizeof (BroadwayRequest));
-
-	      memcpy (&request, buffer, size);
-	      client_handle_request (client, &request);
+	      client_handle_request (client, (BroadwayRequest *)buffer);
 
 	      remaining -= size;
 	      buffer += size;
diff --git a/gdk/broadway/gdkbroadway-server.c b/gdk/broadway/gdkbroadway-server.c
index 2e33469..cb9bdf4 100644
--- a/gdk/broadway/gdkbroadway-server.c
+++ b/gdk/broadway/gdkbroadway-server.c
@@ -550,7 +550,7 @@ create_random_shm (char *name)
 static const cairo_user_data_key_t gdk_broadway_shm_cairo_key;
 
 typedef struct {
-  char name[34];
+  char name[36];
   void *data;
   gsize data_size;
 } BroadwayShmSurfaceData;
@@ -610,7 +610,7 @@ _gdk_broadway_server_window_update (GdkBroadwayServer *server,
   g_assert (data != NULL);
 
   msg.id = id;
-  memcpy (msg.name, data->name, 34);
+  memcpy (msg.name, data->name, 36);
   msg.width = cairo_image_surface_get_width (surface);
   msg.height = cairo_image_surface_get_height (surface);
 
diff --git a/gtk/gtktextview.c b/gtk/gtktextview.c
index 1a0b560..2de8a1e 100644
--- a/gtk/gtktextview.c
+++ b/gtk/gtktextview.c
@@ -9101,7 +9101,8 @@ text_window_scroll        (GtkTextWindow *win,
 
   if (dx != 0 || dy != 0)
     {
-      _gtk_bubble_window_popdown (GTK_BUBBLE_WINDOW (priv->selection_bubble));
+      if (priv->selection_bubble)
+        _gtk_bubble_window_popdown (GTK_BUBBLE_WINDOW (priv->selection_bubble));
       gdk_window_scroll (win->bin_window, dx, dy);
     }
 }
